### PR TITLE
Lit tests for cuda_compile_cubin CMake macro

### PR DESCRIPTION
**Pull request by TejaX-Alaghari**
Adding lit test for cuda_compile_cubin CMake macro

### CODE REVIEW

File: clang/test/dpct/cmake_migation/case_013/case_013.cpp [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// UNSUPPORTED: system-windows
+// RUN: rm -rf %T && mkdir -p %T
+// RUN: cd %T
+// RUN: cp %S/input.cmake ./input.cmake
+// RUN: dpct -in-root ./ -out-root out  ./input.cmake ./vars_def.cmake --rule-file=%S/../../../../tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml --migrate-cmake-script-only --cuda-include-path="%cuda-path/include"
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_013/case_013.cpp]: ./vars_def.cmake is not needed in your case, pls remove it.

File: clang/test/dpct/cmake_migation/case_013/input.cmake [hunk 2]
```diff
@@ -0,0 +1,12 @@
+set(cufile file1.cu)
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_013/input.cmake]: Can file1.cu and file2.cu be built out binaries with input.cmake script?

**Review by tomflinda — CHANGES_REQUESTED** [review]
Pls add the migration yaml based rules for you lit test.

File: clang/test/dpct/cmake_migation/case_013/case_013.cpp (hunk 1 above)
Comment (TejaX-Alaghari) [reply clang/test/dpct/cmake_migation/case_013/case_013.cpp]: I missed to remove that file while copying
Thanks for pointing out 👍

File: clang/test/dpct/cmake_migation/case_013/input.cmake (hunk 2 above)
Comment (TejaX-Alaghari) [reply clang/test/dpct/cmake_migation/case_013/input.cmake]: No, the binaries will not be generated with the input.cmake.
The intention is to get the cubin files (which are generated as below)

Eg.: cubinfiles_1: /5673_CMake_Mig_Lit/cubin/SYCLomatic/clang/test/dpct/cmake_migation/case_013/build/cuda_compile_cubin_3_generated_file1.cu.cubin;

File: clang/test/dpct/cmake_migation/case_013/case_013.cpp (hunk 1 above)
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_013/case_013.cpp]: Line 5 will run fail, and the following checking commands will not be executed.

File: clang/test/dpct/cmake_migation/case_013/expected.txt [hunk 3]
```diff
@@ -0,0 +1,12 @@
+set(cufile file1.cu)
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_013/expected.txt]: Line1 is not an correct expected output, should be "set(cufile file1.dp.cpp)" instead.
Pls make sure lit test can run pass locally before asking for review.

File: clang/test/dpct/cmake_migation/case_013/case_013.cpp (hunk 1 above)
Comment (TejaX-Alaghari) [reply clang/test/dpct/cmake_migation/case_013/case_013.cpp]: Removed the vars_def.cmake arg from above line
The command now runs

**Comment by TejaX-Alaghari** [comment]
> Pls add the migration yaml based rules for you lit test.

Implemented the migration rule

File: clang/test/dpct/cmake_migation/case_013/expected.txt (hunk 3 above)
Comment (TejaX-Alaghari) [reply clang/test/dpct/cmake_migation/case_013/expected.txt]: The expected.txt file is now updated and the test now runs as expected

./ninja_release/./bin/llvm-lit -vv SYCLomatic/clang/test/dpct/cmake_migation/case_013/case_013_cuda_compile_cubin.cpp 
llvm-lit: workspace/JIRA/5673_CMake_Mig_Lit/cubin/SYCLomatic/llvm/utils/lit/lit/llvm/config.py:488: note: using clang: workspace/JIRA/5673_CMake_Mig_Lit/cubin/ninja_release/bin/clang
-- Testing: 1 tests, 1 workers --
PASS: Clang :: dpct/cmake_migation/case_013/case_013_cuda_compile_cubin.cpp (1 of 1)

Testing Time: 0.14s

Total Discovered Tests: 1
  Passed: 1 (100.00%)

File: clang/test/dpct/cmake_migation/case_013/case_013_cuda_compile_cubin.cpp [hunk 4]
```diff
@@ -0,0 +1,11 @@
+// UNSUPPORTED: system-windows
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_013/case_013_cuda_compile_cubin.cpp]: Line1 could be removed, lit test should be also run pass in Window platform.

Comment (TejaX-Alaghari) [reply clang/test/dpct/cmake_migation/case_013/case_013_cuda_compile_cubin.cpp]: Removed the line as suggested

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml [hunk 5]
```diff
@@ -41,3 +41,18 @@
       In: ${arg}.cu
       Out: ${arg}.dp.cpp
 
+- Rule: rule_cuda_compile_cubin
+  Kind: CMakeRule
+  Priority: Fallback
+  CmakeSyntax: cuda_compile_cubin
+  MatchMode: Partial
+  In: cuda_compile_cubin(${value})
```
Comment (tomflinda) [inline clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml]: This line does performs compiling, from a.cu to a.cubin.
So, in SYCL side, the mapping should also perform compiling, from a.dp.cpp to a.dp.o with "-c" option.

File: clang/test/dpct/cmake_migration/case_013/input.cmake [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  ${cufile}
+)
+
+cuda_compile_cubin(cubinfiles_1 file1.cu file2.cu)
+cuda_compile_cubin(cubinfiles_2 
+  file1.cu 
+  file2.cu)
+
+cuda_compile_cubin(cubinfile_1 file1.cu OPTIONS
+  -O3 -Xptxas -v --use_fast_math)
```
Comment (the-slow-one) [inline clang/test/dpct/cmake_migration/case_013/input.cmake:13]: What happens to the `nvcc`'s options here and if they are not supported by the `icpx`?

File: clang/test/dpct/cmake_migation/case_013/input.cmake [hunk 7]
```diff
@@ -0,0 +1,16 @@
+project(foo-bar LANGUAGES CXX CUDA)
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_013/input.cmake]: could be removed.

File: clang/test/dpct/cmake_migation/case_013/input.cmake [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+set(cufile file1.cu)
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_013/input.cmake]: line3 could be removed too.

File: clang/test/dpct/cmake_migation/case_013/expected.txt [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+dpct_compile_sycl_code( cubinfile_2 
+${cufile}
+)
+
+dpct_compile_sycl_code(cubinfiles_1 file1.dp.cpp file2.dp.cpp)
+dpct_compile_sycl_code(cubinfiles_2 file1.dp.cpp 
+file2.dp.cpp)
+
+dpct_compile_sycl_code(cubinfile_1 file1.dp.cpp OPTIONS
+  -O3 -Xptxas -v --use_fast_math)
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_013/expected.txt]: "PTIONS  -O3 -Xptxas -v --use_fast_math" is specific to cuda, not needed to be appeared in SYCL version cmake scripts.

File: clang/test/dpct/cmake_migration/case_013/input.cmake (hunk 6 above)
Comment (TejaX-Alaghari) [reply clang/test/dpct/cmake_migration/case_013/input.cmake:13]: Options will be removed

**Review by tomflinda — APPROVED** [review]
LGTM

**Review by zhiweij1 — APPROVED** [review]

File: clang/test/dpct/cmake_migation/case_013/case_013_cuda_compile_cubin.cpp [hunk 10]
```diff
@@ -0,0 +1,10 @@
+// RUN: rm -rf %T && mkdir -p %T
+// RUN: cd %T
+// RUN: cp %S/input.cmake ./input.cmake
+// RUN: dpct -in-root ./ -out-root out  ./input.cmake --rule-file=%S/../../../../tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml --migrate-cmake-script-only --cuda-include-path="%cuda-path/include"
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_013/case_013_cuda_compile_cubin.cpp]: file=%S/../../../../tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml and  --cuda-include-path="%cuda-path/include" could be removed, just like https://github.com/oneapi-src/SYCLomatic/blob/ba2e8ac690a5a66fac50027935709b2491900fbf/clang/test/dpct/cmake_migation/case_001/case_001_framework.cpp#L4

File: clang/test/dpct/cmake_migation/case_013/expected.txt [hunk 11]
```diff
@@ -0,0 +1,12 @@
+cuda_compile_cubin(cubinfile_1 ${cufile})
```
Comment (zhimingwang36) [inline clang/test/dpct/cmake_migation/case_013/expected.txt]: should line 1, 2 7, 8 be migrated to dpct_compile_sycl_code ?

Comment (TejaX-Alaghari) [reply clang/test/dpct/cmake_migation/case_013/expected.txt]: Yeah, you're right
I think this was updated incorrectly during rebasing

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml [hunk 12]
```diff
@@ -264,3 +264,28 @@
...
+      In: ${arg}.cu
+      Out: ${arg}.dp.cpp
+
+- Rule: rule_cuda_compile_cubin
+  Kind: CMakeRule
+  Priority: Fallback
+  CmakeSyntax: cuda_compile_cubin
+  MatchMode: Partial
+  In: cuda_compile_cubin(${device} ${value} OPTIONS ${opts})
+  Out: dpct_compile_sycl_code(${device} ${value})
```
Comment (tomflinda) [inline clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml]: In the latest code repo, dpct_compile_sycl_code has been renamed to dpct_helper_compile_sycl_code, so you rebase your PR and update the migration rules for cuda_compile_cubin.

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**